File: web/src/api/image.ts
import { post } from "@/utils/request2";

export type AssessAndEnhanceParams = {
  image: string; // base64 字符串（不含 data: 前缀）
  threshold?: number; // 可选的质量阈值
  max_iterations?: number; // 可选的最大增强迭代次数
};

export type AssessAndEnhanceResponse = {
  needs_enhancement: boolean;
  enhanced_image?: string; // base64 编码的增强后图像（不含 data: 前缀）
  overall_score?: number;
  detailed_scores?: Record<string, number>;
  enhancement_iterations?: number;
  success: boolean;
  message?: string;
};

/**
 * 评估图像质量并可选择性增强。
 * POST /pro_pic/assess_and_enhance
 */
export async function assessAndEnhanceImage(
  params: AssessAndEnhanceParams
): Promise<AssessAndEnhanceResponse> {
  const res = await post<AssessAndEnhanceResponse>(
    "/pro_pic/assess_and_enhance",
    params
  );
  return res as AssessAndEnhanceResponse;
}

export default {
  assessAndEnhanceImage,
};

// ===== Pro/Admin: 多变体图片润色 =====
export interface VipOptimizeParams {
  image_base64: string; // 输入图片 base64（不含 data: 前缀）
  num_variants?: number; // 生成变体数量，默认 1
}

export interface VipOptimizeImageItem {
  variant_number: number;
  image_base64: string; // 输出图片 base64（不含 data: 前缀）
  success: boolean;
  error?: string | null;
}

export interface VipOptimizeResponse {
  success: boolean;
  message?: string;
  images: VipOptimizeImageItem[];
  total_generated?: number;
}

/**
 * Pro 专享：一次生成多张优化后的图片变体。
 * POST /pro_pic/vip_optimize
 */
export async function vipOptimizeImages(
  params: VipOptimizeParams
): Promise<VipOptimizeResponse> {
  const res = await post<VipOptimizeResponse>("/pro_pic/vip_optimize", params);
  return res as VipOptimizeResponse;
}

File: py-server/api/routers/image.py
from __future__ import annotations

from typing import Dict
from fastapi import APIRouter, Depends, HTTPException

from schemas.image import ImageRequest, QualityResponse
from core.deps import get_services
from img_pro.utils.image_utils import base64_to_image, image_to_base64, resize_image
from core.settings import get_settings
from core.logging import logger


router = APIRouter(prefix="/pro_pic", tags=["image"])


@router.post("/assess_only", response_model=Dict[str, float])
async def assess_image_quality(request: ImageRequest, services=Depends(get_services)):
    try:
        settings = get_settings()
        image = base64_to_image(request.image)
        image = resize_image(image, settings.MAX_IMAGE_SIZE)
        quality_scores = services.quality_assessor.assess_quality(image)
        return quality_scores
    except Exception as e:
        logger.error(f"Error assessing image quality: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Internal server error: {str(e)}")


@router.post("/assess_and_enhance", response_model=QualityResponse)
async def assess_and_enhance_image(request: ImageRequest, services=Depends(get_services)):
    try:
        settings = get_settings()
        image = base64_to_image(request.image)
        image = resize_image(image, settings.MAX_IMAGE_SIZE)
        quality_scores = services.quality_assessor.assess_quality(image)
        overall_score = quality_scores['overall']
        logger.info(f"Initial quality score: {overall_score:.3f}")

        enhanced_image = None
        iterations = 0
        current_image = image.copy()

        if overall_score < request.threshold:
            logger.info(
                f"Quality below threshold {request.threshold}, starting enhancement..."
            )
            for i in range(request.max_iterations):
                iterations = i + 1
                current_image = services.image_enhancer.enhance_image(current_image, quality_scores)
                new_quality_scores = services.quality_assessor.assess_quality(current_image)
                new_overall_score = new_quality_scores['overall']
                logger.info(
                    f"Enhancement iteration {iterations}: score {new_overall_score:.3f}"
                )
                if new_overall_score >= request.threshold or (new_overall_score - overall_score) < 0.05:
                    quality_scores = new_quality_scores
                    overall_score = new_overall_score
                    break
                quality_scores = new_quality_scores
                overall_score = new_overall_score
            enhanced_image = image_to_base64(current_image)

        return QualityResponse(
            overall_score=overall_score,
            detailed_scores=quality_scores,
            needs_enhancement=overall_score < request.threshold,
            enhanced_image=enhanced_image,
            enhancement_iterations=iterations,
            success=True,
            message=f"Processing completed. Final score: {overall_score:.3f}",
        )
    except Exception as e:
        logger.error(f"Error processing image: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Internal server error: {str(e)}")
